### DISCUSSION TITLE
Adding support for column truncation to 'message-journal'

### DISCUSSION THREAD

**Discussion by subpop** [RedHatInsights/yggdrasil#215] [Ideas]
#116 introduced a message journal and initial support to `yggctl` to retrieve the contents of the journal. The journal output can be filtered using flags that correlate to field names (`--worker`, `--message-id`, `--since`, etc.). The table format currently prints all fields and ends up wrapping the table if the table width exceeds the terminal width. #164 proposes an initial implementation to add the ability to truncate the fields. The flags proposed by this PR support truncating fields by name within the `WORKER DATA` content. Before supporting truncation for worker content, I'd like to consider supporting truncation across the entire table. I looked over the [CLIG](https://clig.dev) and read the section on [command line arguments and flags](https://clig.dev/#arguments-and-flags). It links to some of its source material, which includes an [article on good CLI design decisions](https://medium.com/@jdxcode/12-factor-cli-apps-dd3c227a0e46). This article has a [section on table usage](https://medium.com/@jdxcode/12-factor-cli-apps-dd3c227a0e46#7ee5). This section talks about output formatting, and includes the following:

> Be mindful of the screen width. Only show a few columns by default but allow the user to pass --columns with a comma-separated list of column names to add less common types.
> Truncate rows that are going to spill over the current screen width unless --no-truncate is set.

Could we approach table output support by running down the list of suggested flags in that section? Each flag could be implemented in a separate PR to keep the scope of the PRs under control.

### `--columns`
> Only show a few columns by default but allow the user to pass `--columns` with a comma-separated list of column names to add less common types.

Add a `--columns` flag that accepts a comma-separated list of database table names. Otherwise all the columns are included.

### `--no-truncate`
> Truncate rows that are going to spill over the current screen width unless `--no-truncate` is set.

Add dynamic screen width detection and truncate columns that won't fit based on the screen width. Skip this detection if `--no-truncate` flag is given.

### `--no-headers`
> Show column headers by default but allow them to be hidden with `--no-headers`.

Add `--no-headers` flag that omits a headers row. Headers are visible by default.

### `--filter`
> Allow users to pass `--filter` to filter specific columns. (grep can usually do this, but a flag can filter on specific cell values)

I like the suggestion that grep (or awk) can do this kind of filtering, but allowing the database perform the filter is more performant than text-processing, so I'd like to consider supporting this flag. This would replace the current set of flags we have already: `--message-id`, `--since`, etc. We'd have to figure out a small syntax for parsing the value of this flag. More discussion is needed.

### `--sort`
> Allow sorting by column with `--sort`. Allow inverse and multi-column sort as well.

I don't think we have a the ability to sort the output. Again, we should come up with a simple syntax for this and let the database sort the output.

**Comment by DuckBoss** [comment]
Thanks for writing out your ideas into this discussion thread!
It might be worth it to create a discussion card in Jira to track this.

This is a great start, I've listed some of my thoughts below regarding these flags:

### `--columns`

> Be mindful of the screen width. Only show a few columns by default but allow the user to pass --columns with a comma-separated list of column names to add less common types.
> Truncate rows that are going to spill over the current screen width unless --no-truncate is set.

I have some questions/considerations about how truncation of columns would be handled.
If no options are provided, and columns are truncated when the table width exceeds the terminal width, is there any indication to the user that columns have been truncated? I don't think it's a good idea to not display this info to the user in some way.

Perhaps have the table dimensions printed in the output similar to how Dataframes are handled in python's Pandas library?
Example of a 4 rows by 10 columns table truncated to 4 rows by 4 columns:
```
MESSAGE #   | MESSAGE ID    | SENT            | WORKER NAME     | ...
  0         | <message-id>  | <UTC timestamp> | <worker_name>   | ...
  1         | ...           | ...             | ...             | ...
  2         | ...           | ...             | ...             | ...
  3         | ...           | ...             | ...             | ...

[4 rows x 10 columns]
```

I'm also assuming that `--columns <col1>,<col2>,...` would order the columns as per the user input.

### `--no-truncate`
This sounds like a good change, and the reasoning makes sense.
Is there any benefit to extending truncation features (or any other proposed feature) to non-table outputs?
My previous PR supports it in all output formats (for the worker data column), but I'm not sure if we want that moving forward.

### `--no-headers`
This is a no-brainer for me. Sounds like a good addition.

### `--filter`
Yes! I would love to move the table filtering into a `--filter` option.
We will have to figure out a way to cleanly represent this though.
Maybe something like this, but this is more of an implementation detail...
```bash
$ yggctl message-journal --format table --filter worker:<worker>, message-id:<message-id>
```

### `--sort` 
Is there anything worth sorting except the `SENT` column? 
Every other column that someone would want sorting for would probably benefit more from filtering instead.
Either way this should be a relatively straight-forward addition.

**Reply by subpop** [reply]
> It might be worth it to create a discussion card in Jira to track this.

I thought about this, but I couldn't figure out what the point of a Jira would be. "Please go read this discussion." How do we flow that issue through status?

**Reply by subpop** [reply]
> If no options are provided, and columns are truncated when the table width exceeds the terminal width, is there any indication to the user that columns have been truncated? I don't think it's a good idea to not display this info to the user in some way.

I'm second-guessing this automatic truncation idea. `podman` doesn't truncate the columns of `ps` output; it just wraps. This would certainly be easier than determining terminal width automatically. And if the user wants to reduce the columns included in output, `--columns` is available.

> I'm also assuming that --columns <col1>,<col2>,... would order the columns as per the user input.

Yes! This is a great idea, very intuitive.

**Reply by DuckBoss** [reply]
I believe we had a similar discussion card for something else in Jira, but I don't care for it either way. :+1:
